### ISSUE TITLE
Forking is not possible anymore when using any PyTorch version from about 2 months ago

### ISSUE THREAD

**Issue by Amir-Arsalan** [pytorch/pytorch#20048]
I have been building PyTorch from source in the past couple of months. Since around 1 (or maybe 2) month(s) ago I am unable to do forward passing to my models when using the `multiprocessing` package of PyTorch. If I'm not mistaken, the last version of PyTorch that I built and still allows me to use `fork`ing is [this](https://github.com/pytorch/pytorch/tree/bad4442) one; or more precisely, the version of PyTorch that I am using now and allows me to do `fork`ing is `1.1.0a0+3a01a45`.

Not being able to do `fork`ing is pretty annoying cause I have written all of my work flow in a way that it depends on using `multiprocessing.Process` with `fork`ing. In one part of my work flow I start tens of processes and do forward passing on a model that is on the CPU memory. This is pretty useful and super efficient since the weights of the model are not being copied every time I start a new process. However, if I `spawn` processes the weights of the model needs to get copied and it's not gonna be efficient anymore. 

In another part of my framework, I start a process while the model is still on CPU's memory and then do `model.cuda()` within the process and the model is then copied to the GPU memory. This is a bit inefficient but still allows me to start a 4-5 processes (depending on the GPU memory) and do forward-pass my inputs. 

Here's a pseudocode of what I do:

```
from torch.multiprocessing import Process
model = loadModel() # somehow load a model (e.g. from torch vision)
inputList = loadListOfInputs() # somehow get the list of input tensors

processes = []
for i in range(100):
    processes.append(Process(target=doForwardPass, kwargs={'input': inputList[i]}))
    processes[-1].start()

for i in range(100):
    processes[i].join()

def doForwardPass(input):
    # model.cuda() # this, uncommented, used to work also
    output = model(input)
```

With the current versions of PyTorch (since 1-2 months ago) I cannot do this anymore. I'm afraid this is happening due to some indirect affects of some commits which did not intend to disable `fork`ing. So I wonder if you guys can either revert the changes that have caused this or enable this feature again.

Also, when using some of the PyTorch versions from the master branch (since 1-2 months ago) I might get the following if I do `model.cuda()` within an started process:

`THCudaCheck FAIL file=/pytorch/aten/src/THC/THCGeneral.cpp line=51 error=3 : initialization error`

This might be somewhat relevant to [lazy initialization of CUDA](https://github.com/pytorch/pytorch/pull/2811).

**Comment by SsnL** [comment]
It is surprising that your program ever worked because CUDA runtime does not support `fork`. See this note being added: https://github.com/pytorch/pytorch/pull/19904

**Comment by Amir-Arsalan** [comment]
@SsnL @soumith @ezyang I know CUDA was never `fork`-safe and that's why I didn't do `model.cuda()` before starting any of the processes: cause once CUDA is initialized within one environment/process you cannot use the model in a child environment/process.

What is the solution now? Why can't I even do `fork`ing when the model is on the CPU memory? How come my code always works up until that specific version of PyTorch but not now?

**Comment by soumith** [comment]
there was a bug that was in the last 2 months, but was fixed on master. I am trying to find the relevant issue and fix.

**Comment by soumith** [comment]
I've been searching for the fix, but I think it was titled as something else -- so it's been hard to find. But I can confirm that it was definitely fixed.

**Comment by Amir-Arsalan** [comment]
@soumith Fixed meaning that I should be able to run my code, at least when on CPU? But that's not what's happening now. I cannot use my model in any process.

**Comment by soumith** [comment]
are you saying it doesn't work like that on master, or on a commit from 2 months ago? I thought master is fine (forking before initializing CUDA should work fine)

**Comment by SsnL** [comment]
I think he's saying that the code works 2 months ago, but not anymore on master.

**Comment by Amir-Arsalan** [comment]
@soumith Compiling PyTorch from master (or any commit from the past two month'ish) does not allow me to do `fork`ing. I was able to do `fork`ing on commits from 2 months ago and before that.

**Comment by ezyang** [comment]
So, one interesting thing to note, is that a single call to `cudaGetDeviceCount` is sufficient to initialize CUDA and cause forking to fail. I made some relevant changes during a time when I was not aware of this distinction (#18445), and this might be the cause of the problem.

```
#include <cuda.h>
#include <stdio.h>
#include <unistd.h>

int main() { 
  int c;
  cudaGetDeviceCount(&c);
  fork();
  void* p;
  cudaError_t err = cudaMalloc(&p, 100);
  if (err != cudaSuccess) { 
    printf("failed\n");
  } 
}
```

**Comment by ezyang** [comment]
Update: The particular PR I linked looks blameless, as it doesn't seem to cause cudaGetDeviceCount to be called any more than it was previously. Still investigating.

**Comment by ezyang** [comment]
This script passes on master:

```
import torch
import os

print(torch.zeros(2))

os.fork()

print(torch.zeros(2, device='cuda'))
```

So, at least the "basic" CPU functionality is not broken.

**Comment by ezyang** [comment]
@Amir-Arsalan I "reconstituted" your small script into a running reproducer:

```
import torch
from torch.multiprocessing import Process
import torchvision

model = torchvision.models.resnet50()
inputList = [torch.randn(1, 3, 224, 224), torch.randn(1, 3, 224, 224)]

def doForwardPass(input):
    model.cuda()

processes = []
for i in range(2):
    processes.append(Process(target=doForwardPass, kwargs={'input': inputList[i]}))
    processes[-1].start()

for i in range(2):
    processes[i].join()
```

However, on master, this does not reproduce the problem. Can you dump your entire script here, by any chance? Otherwise, we'll need a more detailed reproducer.

**Comment by Amir-Arsalan** [comment]
@ezyang I'll try to provide a script on Monday.

**Comment by ezyang** [comment]
@Amir-Arsalan I'm going to close this issue for now. Please feel free to reopen when you do have a reproducer.

**Comment by Amir-Arsalan** [comment]
@ezyang Sorry I forgot to work on the example as I got very busy with my stuff. I definitely know that this is an issue and will try to give you an example as soon as possible.

**Comment by ezyang** [comment]
Thanks, looking forward to it.

Excerpts from Amir Arsalan Soltani's message of 2019-06-04 07:06:50 -0700:
> @ezyang Sorry I forgot to work on the example as I got very busy with my stuff. I definitely know that this is an issue and will try to give you an example as soon as possible.
>
